lxd/storage/backend_lxd.go

```diff
@@ -2119,92 +2122,166 @@ func (b *lxdBackend) CreateInstanceFromMigration(inst instance.Instance, conn io
 			}
+
+			snapshotStorageName := project.Instance(inst.Project().Name, instSnapshot.Name())
+			targetSnapshots = append(targetSnapshots, b.GetVolume(volType, contentType, snapshotStorageName, snap.Config))
 		}
-	}
 
-	// Retrieve a list of target volume snapshots.
-	// Afterwards load the volume from the snapshot to ensure the right ordering.
-	instSnapshots, err := inst.Snapshots()
-	if err != nil {
-		return err
-	}
+		volCopy := drivers.NewVolumeCopy(vol, targetSnapshots...)
 
-	targetSnapshots := make([]drivers.Volume, 0, len(instSnapshots))
-	for _, instSnapshot := range instSnapshots {
-		snap, err := VolumeDBGet(b, inst.Project().Name, instSnapshot.Name(), volType)
+		err = b.driver.CreateVolumeFromMigration(volCopy, conn, args, &preFiller, op)
 		if err != nil {
 			return err
 		}
 
-		snapshotStorageName := project.Instance(inst.Project().Name, instSnapshot.Name())
-		targetSnapshots = append(targetSnapshots, b.GetVolume(volType, contentType, snapshotStorageName, snap.Config))
-	}
+		if !isRemoteClusterMove {
+			revert.Add(func() { _ = b.DeleteInstance(inst, op) })
+		}
+	} else {
+		// Get source image path (where image will be uploaded) from volume arguments.
+		srcImagePath := args.ConvertFilePath
 
-	volCopy := drivers.NewVolumeCopy(vol, targetSnapshots...)
+		file, err := os.Create(srcImagePath)
+		if err != nil {
+			return fmt.Errorf("Failed to create temporary file (%s): %w", srcImagePath, err)
+		}
 
-	err = b.driver.CreateVolumeFromMigration(volCopy, conn, args, &preFiller, op)
-	if err != nil {
-		return err
-	}
+		// Remove uploaded image regardless of image conversion success.
+		defer os.Remove(srcImagePath)
+		defer file.Close()
+
+		volCopy := drivers.NewVolumeCopy(vol)
+
+		// Upload image to the backups directory instead of instance volume. This allows us to avoid "doubling"
+		// the quota on instance's volume.
+		err = b.driver.CreateVolumeFromMigration(volCopy, conn, args, nil, op)
+		if err != nil {
+			return err
+		}
 
-	if !isRemoteClusterMove {
 		revert.Add(func() { _ = b.DeleteInstance(inst, op) })
+
+		// Mount volume and get disk path - must be after mount
+		err = b.driver.MountVolume(vol, op)
+		if err != nil {
+			return err
+		}
+
+		revert.Add(func() { _, _ = b.driver.UnmountVolume(vol, false, op) })
+
+		// Extract image format and size.
+		imgFormat, imgBytes, err := qemuImgInfo(srcImagePath)
+		if err != nil {
+			return err
+		}
+
+		// Parse volume size into bytes, so that we can compare them.
+		volBytes, err := units.ParseByteSizeString(vol.ConfigSize())
+		if err != nil {
+			return fmt.Errorf("Failed parsing instance volume size")
+		}
+
+		// Ensure image conversion will fit into the instance volume.
+		if volBytes < imgBytes {
+			// Convert to IEC format for nicer error.
+			imgSize := units.GetByteSizeStringIEC(imgBytes, 2)
+			volSize := units.GetByteSizeStringIEC(volBytes, 2)
+
+			return fmt.Errorf("Volume size (%s) is lower then raw image size (%s)", volSize, imgSize)
+		}
```

Comment:
Would it make sense to also handle the case where the `volBytes` is needlessly bigger than `imgBytes` in which case we could shrink the volume to avoid having unused space at the end?

Reply:
Lets consider this as a separate issue and not expand this task.